### PR TITLE
[AMP] Support XLA:TPU

### PR DESCRIPTION
**Pull request by cowanmeg**
With https://github.com/pytorch/xla/pull/5148, https://github.com/pytorch/xla/pull/4740 

With these changes
XLA:GPU users should use `torch.cuda.amp.autocast()` for AMP with float16
XLA:TPU users should use `torch.amp.autocast('xla')` for AMP with bfloat16

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: c10/core/DispatchKey.h [hunk 1]
```diff
@@ -353,8 +353,8 @@ enum class DispatchKey : uint16_t {
   AutocastCPU,
   AutocastXPU,
   AutocastHPU,
-  // Naughtily, AutocastCUDA is also being used for XLA.  In the terminal state,
-  // it probably should get its own Autocast key
+  AutocastXLA, 
+  // AutocastXLA is only being used for TPUs. XLA GPUs continue to use AutocastCUDA.
```
Comment (bdhirsh) [inline c10/core/DispatchKey.h:357]: @cowanmeg can you describe how this works a bit more - what's the UX here? Is the user expected to use `torch.cuda.autocast()` when using XLA with gpu's, and torch.xla.autocast()` when using tpu's?

File: third_party/kineto
Comment (bdhirsh) [inline third_party/kineto:1]: the submodule updates were probably accidental - don't forget to remove!

File: torch/__init__.py [hunk 2]
```diff
@@ -1364,6 +1364,7 @@ def _assert(condition, message):
 # side effect of adding to the imported module's members for other users.
 from torch import cuda as cuda
 from torch import cpu as cpu
+from torch import xla as xla
```
Comment (bdhirsh) [inline torch/__init__.py:1367]: I don't think we want this - why do you need it? This will have `torch.xla` show up by default as an available module when you `import torch`. But that doesn't really make sense for anyone not using `torch_xla`.

File: torch/csrc/jit/passes/autocast.cpp [hunk 3]
```diff
@@ -24,11 +24,13 @@ bool autocast_enabled = true;
 struct AutocastContext {
   bool gpu_enabled = false;
   bool cpu_enabled = false;
+  bool xla_enabled = false;
```
Comment (bdhirsh) [inline torch/csrc/jit/passes/autocast.cpp:27]: cc @davidberard98 - do you mind reviewing the JIT changes in this file? I'm not too familiar with them.

File: torch/csrc/jit/runtime/static/generated_ops.cpp [hunk 4]
```diff
@@ -5191,10 +5191,12 @@ REGISTER_NATIVE_OPERATOR_FUNCTOR(
           const auto& self = p_node->Input(0).toTensor();
           const auto cuda_enabled = p_node->Input(1).toBool();
           const auto cpu_enabled = p_node->Input(2).toBool();
-          const auto cuda_dtype = p_node->Input(3).toScalarType();
-          const auto cpu_dtype = p_node->Input(4).toScalarType();
+          const auto xla_enabled = p_node->Input(3).toBool();
```
Comment (davidberard98) [inline torch/csrc/jit/runtime/static/generated_ops.cpp:5194]: @tenpercent can you take a look at this? should we just leave static runtime out?

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -7129,11 +7129,11 @@
 - func: choose_qparams_optimized(Tensor input, int numel, int n_bins, float ratio, int bit_width) -> (Tensor, Tensor)
   variants: function
 
-- func: _autocast_to_reduced_precision(Tensor(a) self, bool cuda_enabled, bool cpu_enabled, ScalarType cuda_dtype, ScalarType cpu_dtype) -> Tensor(a)
+- func: _autocast_to_reduced_precision(Tensor(a) self, bool cuda_enabled, bool cpu_enabled, bool xla_enabled, ScalarType cuda_dtype, ScalarType cpu_dtype, ScalarType xla_dtype) -> Tensor(a)
```
Comment (davidberard98) [inline aten/src/ATen/native/native_functions.yaml:7132]: I think this is bc-breaking? not very familiar with how to do this, but I think we'd need to add upgraders for jit, right?

File: torch/csrc/jit/passes/autocast.cpp (hunk 3 above)
Comment (davidberard98) [reply torch/csrc/jit/passes/autocast.cpp:27]: looks good other than the two other comments (static runtime & bc-breaking)

File: aten/src/ATen/autocast_mode.cpp [hunk 6]
```diff
@@ -503,6 +530,144 @@ TORCH_LIBRARY_IMPL(aten, Autocast, m) {
          TORCH_FN((&at::autocast::binary_cross_entropy_banned)));
 }
 
+
+TORCH_LIBRARY_IMPL(_, AutocastXLA, m) {
+  m.fallback(torch::CppFunction::makeFallthrough());
+}
```
[the remark this answers is not in the document]
Comment (cowanmeg) [reply aten/src/ATen/autocast_mode.cpp]: Moved into pytorch/xla. Note, I moved the CastPolicy enum into autocast_mode.h so it could be included.

File: c10/core/DispatchKey.h (hunk 1 above)
Comment (cowanmeg) [reply c10/core/DispatchKey.h:357]: Correct. Updated the summary for clarity.

File: aten/src/ATen/autocast_mode.h [hunk 7]
```diff
@@ -44,13 +48,35 @@ bool is_autocast_eligible(const Tensor& tensor, DeviceType device_type) {
...
       return tensor.device().type() == DeviceType::PrivateUse1 &&
           tensor.is_floating_point();
     default:
       return false;
   }
 }
+
+// Policies correspond to op categories that need code-divergent handling.
+// Wrapper templates below are specialized based on a policy template parameter.
+enum class CastPolicy : uint8_t {
```
Comment (bdhirsh) [inline aten/src/ATen/autocast_mode.h:63]: Why was this moved to a (public) header file?

File: torch/xla/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+from . import amp
```
Comment (bdhirsh) [inline torch/xla/__init__.py:1]: I don't think these `torch/xla/*` files should live in core. instead, any `torch.xla` autocast related API's can live in the pytorch/xla repo, and can be installed when you import `torch_xla`.

**Comment by bdhirsh** [comment]
> XLA:GPU users should use torch.cuda.amp.autocast() for AMP with float16
XLA:TPU users should use torch.xla.amp.autocast() for AMP with bfloat16

@cowanmeg today, there's a `torch.autocast(device=...)` API that is meant to be device generic. Instead of adding a dedicated API namespace (`torch.xla.amp.autocast`), can we instead just get `torch.autocast(device='xla')` to work and recommend that path?

File: aten/src/ATen/autocast_mode.h (hunk 7 above)
Comment (cowanmeg) [reply aten/src/ATen/autocast_mode.h:63]: I moved this into the public header file, so we could import the same CastPolicy in Pytorch/XLA. Alternatively, we could just copy it PyTorch/XLA repo to minimize changes here.

**Comment by cowanmeg** [comment]
> 

Makes sense! Thanks for the feedback.

File: aten/src/ATen/autocast_mode.h (hunk 7 above)
Comment (bdhirsh) [reply aten/src/ATen/autocast_mode.h:63]: ah ok seems fine :) Just wanted to make sure it was intentional.

File: torch/xla/__init__.py (hunk 8 above)
Comment (cowanmeg) [reply torch/xla/__init__.py]: xla/* files have been removed.

**Comment by bdhirsh** [comment]
@cowanmeg out of curiosity - could you elaborate on the differences between autocast rules for TPU and GPU (and what causes that difference)?

**Comment by cowanmeg** [comment]
> @cowanmeg out of curiosity - could you elaborate on the differences between autocast rules for TPU and GPU (and what causes that difference)?

The TPU autocast rules use bfloat16 for lower precision vs float16 for the GPU rules.

In XLATensor, when the backend device is known, the AutocastXLA dispatch key is swapped for AutocastCUDA when a the backend device is a GPU, so that the GPU autocast rules are applied. 
https://github.com/pytorch/xla/blob/aabbd128b9f168c3388b5ad054d99b7734fb5fe0/torch_xla/csrc/tensor_impl.cpp#L66

**Review by bdhirsh — APPROVED** [review]

**Comment by cowanmeg** [comment]
@bdhirsh, the CI errors have been fixed. Can we try merging again?

**Comment by bdhirsh** [comment]
Hey @cowanmeg, looks like the XLA CI is still failing: https://github.com/pytorch/pytorch/actions/runs/4951204871/jobs/8859359749?pr=96370

**Comment by cowanmeg** [comment]
Failure is related to an earlier xla pin update: https://github.com/pytorch/pytorch/pull/100369 
We will try again once we fix that error

**Review by malfet — APPROVED** [review]

**Comment by cowanmeg** [comment]
I think this failing inductor test is unrelated: 

```
2023-06-20T21:08:10.1240937Z [2023-06-20 21:08:10,123] torch._dynamo.utils: [ERROR] RMSE (res-fp64): 0.00213, (ref-fp64): 0.00064 and shape=torch.Size([256])
2023-06-20T21:08:10.1245080Z [2023-06-20 21:08:10,123] torch._dynamo.utils: [ERROR] Accuracy failed for key name backbone.fpn.layer_blocks.2.0.bias.grad
2023-06-20T21:08:10.1296323Z fail_accuracy
```

**Comment by bdhirsh** [comment]
@cowanmeg hmm, I don't think I see that failure in CI on the main branch https://hud.pytorch.org/, and it's a bit hard to tell immediately if it's flaky/unrelated, since that tests E2E logic, and appears to be running with autocast enabled. Can you try rebasing?

**Comment by cowanmeg** [comment]
I think the BC lint failure is a cancellation?
